### PR TITLE
[ALLUXIO-3227] Refactor waitfor

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-3227

- Change the waitFor signatures to all take `Supplier<Boolean>` instead of `Function<Void, Boolean>`
- Change waitFor to declare throwing TimeoutException and InterruptedException.
- implement `waitFor` using `waitForResult` so that they can share the same core implementation
- Change call sites to use lambdas and handle TimeoutException/InterruptedException if necessary
- Add a few unit tests for `waitFor`

The only part that requires a close review is the change to `CommonUtils.java`. The refactors of the call sites are mostly mechanical.

### CODE REVIEW

**Comment by aaudiber** [comment]
alluxio-bot, check this please

File: core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java [hunk 1]
```diff
@@ -120,17 +120,21 @@ public void stop() {
...
           return getExecutorService().awaitTermination(100, TimeUnit.MILLISECONDS);
         } catch (InterruptedException e) {
           throw new RuntimeException(e);
         }
-      }
-    });
+      });
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      return;
```
Comment (apc999) [inline core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java]: do we need to throw RTE here?

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 2]
```diff
@@ -257,17 +257,20 @@ public void stop() {
...
         try {
           return getExecutorService().awaitTermination(100, TimeUnit.MILLISECONDS);
         } catch (InterruptedException e) {
           throw new RuntimeException(e);
         }
-      }
-    });
+      });
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```
Comment (apc999) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: RTE here?

File: tests/src/test/java/alluxio/client/fs/LoadMetadataIntegrationTest.java [hunk 3]
```diff
@@ -254,35 +253,29 @@ private void checkGetStatus(final String path, GetStatusOptions options, boolean
...
-              if (cache.isAbsent(new AlluxioURI(path))) {
-                return true;
-              }
-              return false;
-            }
-          }, WaitForOptions.defaults().setTimeoutMs(60000));
+              .getMaster(FileSystemMaster.class),
+          "mUfsAbsentPathCache");
+      CommonUtils.waitFor("path (" + path + ") to be added to absent cache", () -> {
+        if (cache.isAbsent(new AlluxioURI(path))) {
```
[the remark this answers is not in the document]
Comment (aaudiber) [reply tests/src/test/java/alluxio/client/fs/LoadMetadataIntegrationTest.java]: updated

File: core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java (hunk 1 above)
Comment (aaudiber) [reply core/server/worker/src/main/java/alluxio/worker/file/DefaultFileSystemWorker.java]: updated

File: shell/src/main/java/alluxio/cli/fs/command/FreeCommand.java [hunk 4]
```diff
@@ -71,19 +72,23 @@ protected void runPlainPath(AlluxioURI path, CommandLine cl)
...
-      } catch (Exception e) {
-        Throwables.propagateIfPossible(e);
-        throw new RuntimeException(e);
-      }
-    }, WaitForOptions.defaults().setTimeoutMs(10 * Math.toIntExact(interval))
-        .setInterval(interval));
+      }, WaitForOptions.defaults().setTimeoutMs(10 * Math.toIntExact(interval))
+          .setInterval(interval));
+    } catch (TimeoutException | InterruptedException e) {
+      throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (aaudiber) [reply shell/src/main/java/alluxio/cli/fs/command/FreeCommand.java]: updated

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java (hunk 2 above)
Comment (aaudiber) [reply core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: updated

**Pull request merged**